### PR TITLE
dfe-analytics-dataform v1.8.2

### PR DESCRIPTION
**Pull request by stevenleggdfe**
- Fix behaviour of parameters to turn off data freshness assertion during date ranges. They didn't work because the JS was creating new Date objects set to the current timestamp all over the place instead of processing the specified dates as intended.
- Accelerate checksum performance and exclude/count rows which may have changed since checksum was calculated to match dfe-analytics logic better.
- Add checks that all parameters passed to ```dfeAnalyticsDataform()``` are valid
- Refactor parameter logic into separate module and break into functions to enable unit testing in future

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -99,19 +99,26 @@ module.exports = (params) => {
...
+ const rightNow = new Date();
+
   // Check format of all ranges in the assertionDisableDuringDateRanges array of range objects, and convert month-day formats to date ranges that start within the current year
   params.assertionDisableDuringDateRanges.forEach(range => {
     if (range.fromDate instanceof Date && range.toDate instanceof Date) {
       // range is between two dates already so no further processing required
     }
     else if (Number.isInteger(range.fromMonth) && Number.isInteger(range.fromDay) && Number.isInteger(range.toMonth) && Number.isInteger(range.toDay)) {
-      range.fromDate = Date(Date.now().getFullYear, range.fromMonth, range.fromDay);
-      range.toDate = Date(Date.now().getFullYear, range.toMonth, range.toDay);
```
Comment (ericaporter) [inline index.js:109]: I think you could do something like 'range.fromDate = new Date(rightNow.getFullYear(), range.fromMonth -1, range.fromDay);

Comment (ericaporter) [inline index.js:109]: I think you could do similar to above e.g.: `range.toDate = new Date(rightNow.getFullYear(), range.toMonth - 1, range.toDay);`

File: includes/parameter_functions.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        if (!tableSchema.materialisation) {
+            tableSchema.materialisation = 'table';
+        }
+    });
+    return params;
+}
+
+function dateRangesToDisableAssertionsNow(ranges, currentDate) {
+    let disableAssertionsNow = false;
+    /*let dateRanges = JSON.parse(JSON.stringify(ranges));*/
```
Comment (ericaporter) [inline includes/parameter_functions.js:95]: We could just delete this line

File: index.js [hunk 3]
```diff
@@ -18,145 +20,91 @@ const dataSchemaJSONLatest = require("./includes/data_schema_json_latest");
...
+            fromDay: 29,
+            toMonth: 4,
+            toDay: 14
+        }, {
+            fromMonth: 12,
+            fromDay: 22,
+            toMonth: 1,
+            toDay: 7
+        }], // an array of day or date ranges between which some assertions will be disabled if other parameters are set to disable them. Each range is a hash containing either the integer values fromDay, fromMonth, toDay and toMonth *or* the date values fromDate and toDate. Defaults to an approximation to school holidays each year.
+        ...params
```
Comment (ericaporter) [inline index.js:61]: Out of interest will these be updated each year?

**Review by ericaporter — APPROVED** [review]
Left two comments but they are incidental - happy to approve 👍

File: index.js (hunk 3 above)
Comment (stevenleggdfe) [reply index.js:61]: We could do I guess? Although term dates are pretty loosely defined anyway (different LAs and MATs run to different terms, and bank holidays and inset days fall in random places) so probably not much value in that.

Teams can always override the defaults if they want to.

Also I hope to dispense with data freshness assertions entirely if we manage to get to reliable enough data load such that the checksums can take their place.

**Pull request merged**